src/domain/messages/helpers/message-verifier.helper.ts

```diff
@@ -0,0 +1,136 @@
+
+  constructor(
+    @Inject(IConfigurationService)
+    private readonly configurationService: IConfigurationService,
+  ) {
+    this.isMessageVerificationEnabled = this.configurationService.getOrThrow(
+      'features.messageVerification',
+    );
+  }
+
+  public async verifyCreation(args: {
+    chainId: string;
+    safe: Safe;
+    message: string | Record<string, unknown>;
+    signature: `0x${string}`;
+  }): Promise<void> {
+    if (!this.isMessageVerificationEnabled) {
+      return;
+    }
+
+    // We can't verify the messageHash as we have no comparison hash
+
+    let calculatedHash: `0x${string}`;
+    try {
+      calculatedHash = getSafeMessageMessageHash({
+        chainId: args.chainId,
+        safe: args.safe,
+        message: args.message,
+      });
+    } catch {
+      throw new UnprocessableEntityException('Could not calculate messageHash');
+    }
+
+    await this.verifySignature({
+      safe: args.safe,
+      chainId: args.chainId,
+      messageHash: calculatedHash,
+      signature: args.signature,
+    });
+  }
+
+  public async verifyUpdate(args: {
+    chainId: string;
+    safe: Safe;
+    message: Message;
+    messageHash: `0x${string}`;
+    signature: `0x${string}`;
+  }): Promise<void> {
+    if (!this.isMessageVerificationEnabled) {
+      return;
+    }
+
+    this.verifyMessageHash({
+      chainId: args.chainId,
+      safe: args.safe,
+      message: args.message.message,
+      expectedHash: args.messageHash,
+    });
+
+    await this.verifySignature({
+      safe: args.safe,
+      chainId: args.chainId,
+      messageHash: args.messageHash,
+      signature: args.signature,
+    });
+  }
+
+  public verifyMessageHash(args: {
+    chainId: string;
+    safe: Safe;
+    expectedHash: `0x${string}`;
+    message: string | Record<string, unknown>;
+  }): void {
+    let calculatedHash;
+    try {
+      calculatedHash = getSafeMessageMessageHash({
+        chainId: args.chainId,
+        safe: args.safe,
+        message: args.message as string | TypedDataDefinition,
+      });
+    } catch {
+      throw new BadGatewayException('Could not calculate messageHash');
+    }
+
+    if (calculatedHash !== args.expectedHash) {
+      throw new BadGatewayException('Invalid messageHash');
+    }
+  }
+
+  public async verifySignature(args: {
+    safe: Safe;
+    chainId: string;
+    messageHash: `0x${string}`;
+    signature: `0x${string}`;
+  }): Promise<void> {
+    const signerAddress = await this.recoverSignerAddress({
+      messageHash: args.messageHash,
+      signature: args.signature,
+    });
+
+    const isOwner = args.safe.owners.includes(signerAddress);
+    if (!isOwner) {
+      throw new BadGatewayException('Invalid signature');
+    }
+  }
+
+  public async recoverSignerAddress(args: {
+    messageHash: `0x${string}`;
+    signature: `0x${string}`;
+  }): Promise<`0x${string}`> {
+    // TODO: Throw is v is eth_sign and feature disabled
```

Comment:
We should create a separate signature helper and reuse the logic from the transaction verifier. Let's do it in a follow up though.